db/coordination.go

```diff
@@ -5,115 +5,207 @@ import (
 )
 
 type BothDatabases struct {
-	dedb   *DEDatabase
-	icatdb *ICATDatabase
+	deconn        DatabaseTxAccessor
+	icatconn      DatabaseTxAccessor
+	configuration *config.Config
+
+	DERollback func()
+	DECommit   func() error
+
+	ICATRollback func()
+	ICATCommit   func() error
+
+	detx   *DEDatabase
+	icattx *ICATDatabase
 }
 
-func NewBoth(dedb *DEDatabase, icatdb *ICATDatabase) *BothDatabases {
-	return &BothDatabases{dedb: dedb, icatdb: icatdb}
+func NewBoth(dedb DatabaseTxAccessor, icatdb DatabaseTxAccessor, config *config.Config) *BothDatabases {
+	return &BothDatabases{deconn: dedb, icatconn: icatdb, configuration: config}
 }
 
-func NewBothTx(context context.Context, deconn *sqlx.DB, deschema string, icatconn *sqlx.DB, userSuffix, zone string, rootResourceNames []string) (*BothDatabases, func(), func() error, error) {
-	logStats("DE", deconn)
-	detx, err := deconn.BeginTxx(context, nil)
-	if err != nil {
-		return nil, func() {}, func() error { return err }, errors.Wrap(err, "Error creating DE database transaction")
+func (b *BothDatabases) DETx(ctx context.Context) (*DEDatabase, error) {
+	logStats("DE", b.deconn)
+	if b.detx != nil {
+		return b.detx, nil
 	}
 
-	logStats("ICAT", icatconn)
-	icattx, err := icatconn.BeginTxx(context, nil)
+	detx, err := b.deconn.BeginTxx(ctx, nil)
 	if err != nil {
-		return nil, func() {}, func() error { return err }, errors.Wrap(err, "Error creating ICAT transaction")
+		return nil, errors.Wrap(err, "Error creating DE transaction")
 	}
 
-	dedb := NewDE(detx, deschema)
-	icatdb := NewICAT(icattx, userSuffix, zone, rootResourceNames)
-
 	rb := func() {
 		alreadyDoneErr := "sql: transaction has already been committed or rolled back"
 		err := detx.Rollback()
 		if err != nil && err.Error() != alreadyDoneErr {
 			e := errors.Wrap(err, "Error rolling back DE database transaction")
 			log.Error(e)
 		}
-		err = icattx.Rollback()
-		if err != nil && err.Error() != alreadyDoneErr {
-			e := errors.Wrap(err, "Error rolling back ICAT transaction")
-			log.Error(e)
-		}
+		b.detx = nil
+		b.DERollback = func() {}
+		b.DECommit = func() error { return nil }
 	}
 
 	commit := func() error {
 		err := detx.Commit()
+		b.detx = nil
+		b.DERollback = func() {}
+		b.DECommit = func() error { return nil }
 		if err != nil {
 			e := errors.Wrap(err, "Error committing DE database transaction")
 			log.Error(e)
-			icattx.Rollback() // nolint:errcheck
 			return e
 		}
-		err = icattx.Commit()
+		return nil
+	}
+
+	b.detx = NewDE(detx, b.configuration)
+	b.DERollback = rb
+	b.DECommit = commit
+
+	return b.detx, nil
+}
+
+func (b *BothDatabases) ICATTx(ctx context.Context) (*ICATDatabase, error) {
+	logStats("ICAT", b.icatconn)
+	if b.icattx != nil {
+		return b.icattx, nil
+	}
+
+	icattx, err := b.icatconn.BeginTxx(ctx, nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error creating DE transaction")
+	}
+
+	rb := func() {
+		alreadyDoneErr := "sql: transaction has already been committed or rolled back"
```

Comment:
I only check for these `committed or rolled back` errors (and ignore them) in the rollbacks. I figure those are what get called with `defer` and so are more likely to get called an extra time, and we likely actually care if it happens in the commit.